File: src/deadlock_build_sync/offline/beam_nomination.py
"""Freeze supported beam proposals before validation outcomes are evaluated."""

from __future__ import annotations

from dataclasses import dataclass, replace
from typing import TYPE_CHECKING

from deadlock_build_sync.build_support import SUPPORT
from deadlock_build_sync.mechanics import InventoryState, purchase_item
from deadlock_build_sync.value_validation import (
    integer,
    require_object_dict,
    require_object_list,
    require_object_rows,
)

from .beam_model import BeamItemModel, load_beam_model
from .beam_search import BeamRoute, GroupSearchRequest, search_group
from .beam_support import BeamOwnership
from .core_discovery import calculate_core_identity
from .discovery_artifacts import freeze_purchase_guide
from .discovery_data import load_hero_discovery_data, prepare_discovery_partitions
from .discovery_export import _open_discovery_database
from .discovery_orders import calculate_order_evidence, select_core_purchase_times
from .discovery_quality import evaluate_core
from .discovery_tactics import describe_mechanic_overlap
from .discovery_types import NominatedCoreBuild, SelectedPurchaseOrder

if TYPE_CHECKING:
    import duckdb

    from deadlock_build_sync.mechanics import ItemGraph

    from .discovery_data import HeroDiscoveryData
    from .production_sources import _HeroExportContext


@dataclass
class BeamNomination:
    group: str
    row: NominatedCoreBuild
    scores: dict[str, float]


@dataclass(frozen=True)
class BeamDiscoveryJob:
    hero: dict[str, object]
    baseline: dict[str, object]
    context: _HeroExportContext
    order_only: bool = False


@dataclass
class BeamHeroProposals:
    proposals: list[BeamNomination]
    diagnostics: list[dict[str, object]]


def baseline_groups(baseline: dict[str, object]) -> dict[str, list[dict[str, object]]]:
    groups: dict[str, list[dict[str, object]]] = {}
    for build in require_object_rows(baseline["builds"]):
        groups.setdefault(str(build["guide_group_id"]), []).append(build)
    return dict(
        sorted(
            groups.items(),
            key=lambda pair: min(
                integer(require_object_dict(build["discovery"])["selection_rank"])
                for build in pair[1]
            ),
        )
    )


def core_items(build: dict[str, object]) -> frozenset[int]:
    policy = require_object_dict(build["core_policy"])
    return frozenset(
        integer(item) for item in require_object_list(policy["default_item_ids"])
    )


def beam_purchase_order(
    values: HeroDiscoveryData, route: BeamRoute, graph: ItemGraph
) -> SelectedPurchaseOrder:
    order = [item for item in route.targets if item in route.core]
    evidence = {
        fold: calculate_order_evidence(
            select_core_purchase_times(values, list(route.core), fold),
            list(route.core),
            order,
        )
        for fold in ("discovery", "selection")
    }
    inventory = InventoryState()
    actions: list[dict[str, object]] = []
    spent = 0
    for item in route.path:
        credit = graph.credited_component_value(item, inventory.owned)
        cash = graph.incremental_cash_cost(item, inventory.owned)
        inventory = purchase_item(graph, inventory, item)
        spent += cash
        actions.append({
            "item_id": item,
            "name": graph.require(item).name,
            "role": "core" if item in route.core else "required_component",
            "component_credit": credit,
            "incremental_cost": cash,
            "cumulative_cost": spent,
            "owned_after": list(inventory.owned),
        })
    if set(inventory.owned) != set(route.core) or spent != route.spent:
        raise ValueError("Beam route differs from production mechanics replay")
    admitted = all(row["passes"] for row in evidence.values())
    return {
        "method": "beam16",
        "order": order,
        "discovery": evidence["discovery"],
        "selection": evidence["selection"],
        "admitted_before_validation": admitted,
        "legal": True,
        "actions": actions,
        "reason": None
        if admitted
        else "Beam order lacks discovery or selection support",
    }


def nominate_route(
    connection: duckdb.DuckDBPyConnection,
    job: BeamDiscoveryJob,
    values: HeroDiscoveryData,
    route: BeamRoute,
) -> tuple[NominatedCoreBuild | None, str | None]:
    graph = job.context.item_graph
    path = beam_purchase_order(values, route, graph)
    if not path["admitted_before_validation"]:
        return None, path["reason"]
    discovery = evaluate_core(values, route.core, "discovery")
    selection = evaluate_core(values, route.core, "selection")
    reasons = SUPPORT.core_reasons(discovery["owners"], selection["owners"])
    if reasons:
        return None, "; ".join(reasons)
    row: NominatedCoreBuild = {
        "items": list(route.core),
        "names": [graph.require(item).name for item in route.core],
        "identity_id": calculate_core_identity(values.hero, list(route.core)),
        "cost": sum(graph.require(item).cost for item in route.core),
        "score": route.score,
        "discovery_support": discovery["owners"],
        "discovery_lift": discovery["joint_lift"],
        "selection": selection,
        "selection_rejections": [],
        "selection_rank": 0,
        "hero_id": values.hero,
        "path": path,
        "branch_candidates": [],
        "tactics": describe_mechanic_overlap(
            job.hero, list(route.core), job.context.normal_assets
        ),
    }
    row["guide"] = freeze_purchase_guide(
        connection, values, row, graph, exact_path=route.path
    )
    if not row["guide"]["ready"]:
        return None, row["guide"].get("reason", "Incomplete beam guide")
    return row, None


def nominate_search(
    connection: duckdb.DuckDBPyConnection,
    job: BeamDiscoveryJob,
    values: HeroDiscoveryData,
    request: GroupSearchRequest,
    default: dict[str, object],
) -> BeamHeroProposals:
    if request.group is None:
        raise ValueError("Grouped nomination requires a group identifier")
    result = search_group(request)
    rejected: dict[str, int] = {}
    proposals: list[BeamNomination] = []
    accepted: set[tuple[int, ...]] = set()
    routes = sorted(
        result.routes,
        key=lambda route: (
            -route.score,
            frozenset(route.core) != core_items(default),
            -route.owners,
            route.core,
            route.path,
        ),
    )
    for route in routes:
        if route.core in accepted:
            continue
        row, reason = nominate_route(connection, job, values, route)
        if row is None:
            label = reason or "Unsupported beam route"
            rejected[label] = rejected.get(label, 0) + 1
            continue
        accepted.add(route.core)
        proposals.append(
            BeamNomination(request.group, row, {str(request.state): route.score})
        )
    return BeamHeroProposals(
        proposals,
        [
            {
                "group": request.group,
                "state": request.state,
                "routes": len(result.routes),
                "admitted_cores": len(accepted),
                "rejections": rejected,
                "unassigned": result.unassigned,
                "expansions": result.expansions,
                "search_seconds": result.elapsed_seconds,
            }
        ],
    )


def search_requests(
    job: BeamDiscoveryJob, values: HeroDiscoveryData, model: BeamItemModel
) -> list[tuple[GroupSearchRequest, dict[str, object]]]:
    grouped = baseline_groups(job.baseline)
    groups = {
        group: tuple(core_items(build) for build in builds)
        for group, builds in grouped.items()
    }
    requests = []
    for state in (1, 0, 2):
        ownership = BeamOwnership(values, job.context.item_graph, state)
        for group, builds in grouped.items():
            default = next(build for build in builds if build["path_id"] == group)
            request = GroupSearchRequest(
                job.context.item_graph,
                model,
                ownership,
                groups,
                group,
                state,
                values.items,
            )
            requests.extend(
                (candidate, default)
                for candidate in group_requests(
                    request, builds, order_only=job.order_only
                )
            )
    return requests


def group_requests(
    request: GroupSearchRequest, builds: list[dict[str, object]], *, order_only: bool
) -> list[GroupSearchRequest]:
    if not order_only:
        return [request]
    return [
        replace(
            request,
            candidates=tuple(sorted(core_items(build))),
            required_core=core_items(build),
        )
        for build in builds
    ]


def discover_beam_hero(job: BeamDiscoveryJob) -> BeamHeroProposals:
    cohort = require_object_dict(job.baseline["cohort"])
    minimum, maximum = (
        integer(cohort["minimum_badge"]),
        integer(cohort["maximum_badge"]),
    )
    proposals: dict[tuple[str, tuple[int, ...], tuple[int, ...]], BeamNomination] = {}
    diagnostics: list[dict[str, object]] = []
    with _open_discovery_database(job.context) as connection:
        prepare_discovery_partitions(connection)
        values = load_hero_discovery_data(
            connection,
            integer(job.hero["id"]),
            job.context.item_graph,
            minimum,
            maximum,
        )
        model = load_beam_model(connection, values, minimum, maximum)
        for request, default in search_requests(job, values, model):
            result = nominate_search(connection, job, values, request, default)
            diagnostics.extend(result.diagnostics)
            for proposed in result.proposals:
                key = (
                    proposed.group,
                    tuple(proposed.row["items"]),
                    tuple(proposed.row["guide"]["path"]),
                )
                nomination = proposals.setdefault(key, proposed)
                nomination.scores.update(proposed.scores)
    print(f"Beam proposals: {job.hero['name']} ({len(proposals)} routes)", flush=True)
    return BeamHeroProposals(list(proposals.values()), diagnostics)

File: crates/deadlock-analysis/src/beam_nomination.rs
use std::collections::{BTreeMap, BTreeSet};

use deadlock_data::{Error, Result, array, integer};
use deadlock_guides::SUPPORT;
use serde::{Deserialize, Serialize};
use serde_json::{Value, json};

use crate::beam_model::BeamModel;
use crate::beam_search::{BeamRoute, SearchRequest, search_group};
use crate::beam_support::{BeamOwnership, CoreGroups};
use crate::config::cohort_ranks;
use crate::core_discovery::core_identity;
use crate::core_outcomes::evaluate_core;
use crate::database::AnalysisDatabase;
use crate::discovery_data::{DiscoveryData, load_discovery_data};
use crate::discovery_models::{ExportContext, Nomination, item_ids};
use crate::mechanic_overlap::describe_overlap;
use crate::mining::Candidate;
use crate::purchase_orders::{order_evidence, replay_actions};
use crate::purchase_pool::freeze_guide;

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct BeamNomination {
    pub group: String,
    pub row: Nomination,
    pub scores: BTreeMap<String, f64>,
}

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct HeroProposals {
    pub proposals: Vec<BeamNomination>,
    pub diagnostics: Vec<Value>,
}

pub fn core_items(build: &Value) -> Result<BTreeSet<u64>> {
    Ok(item_ids(&build["core_policy"]["default_item_ids"])?
        .into_iter()
        .collect())
}

pub fn baseline_groups(baseline: &Value) -> Result<Vec<(String, Vec<Value>)>> {
    let mut groups = BTreeMap::<String, Vec<Value>>::new();
    for build in array(&baseline["builds"])? {
        groups
            .entry(deadlock_data::text(build, "guide_group_id")?.into())
            .or_default()
            .push(build.clone());
    }
    let mut groups = groups.into_iter().collect::<Vec<_>>();
    groups.sort_by_key(|(_, builds)| {
        builds
            .iter()
            .filter_map(|build| build["discovery"]["selection_rank"].as_u64())
            .min()
    });
    Ok(groups)
}

pub fn discover_beam_hero(
    hero: &Value,
    baseline: &Value,
    context: &ExportContext,
) -> Result<HeroProposals> {
    let identifier = integer(hero, "id")?;
    let ranks = cohort_ranks(&baseline["cohort"])?;
    let database = context.open_database(identifier)?;
    let data = load_discovery_data(&database, identifier, &context.graph, ranks)?;
    let model = BeamModel::load(&database, &data, ranks)?;
    let grouped = baseline_groups(baseline)?;
    let groups = grouped
        .iter()
        .map(|(name, builds)| {
            Ok((
                name.clone(),
                builds.iter().map(core_items).collect::<Result<Vec<_>>>()?,
            ))
        })
        .collect::<Result<CoreGroups>>()?;
    let mut proposals = BTreeMap::<(String, Vec<u64>, Vec<u64>), BeamNomination>::new();
    let mut diagnostics = Vec::new();
    for state in [1, 0, 2] {
        let ownership = BeamOwnership::new(&data, &context.graph, state)?;
        for (group, builds) in &grouped {
            let default = builds
                .iter()
                .find(|build| build["path_id"] == *group)
                .ok_or_else(|| Error::new("Beam group has no default build"))?;
            let request = SearchRequest {
                graph: &context.graph,
                model: &model,
                ownership: &ownership,
                groups: &groups,
                group,
                state,
                candidates: &data.items,
            };
            let result = nominate_search(&database, hero, &data, &request, default, context)?;
            diagnostics.extend(result.diagnostics);
            for proposed in result.proposals {
                let key = (
                    proposed.group.clone(),
                    proposed.row.candidate.items.clone(),
                    proposed.row.guide.path.clone(),
                );
                if let Some(previous) = proposals.get_mut(&key) {
                    previous.scores.extend(proposed.scores);
                } else {
                    proposals.insert(key, proposed);
                }
            }
        }
    }
    eprintln!(
        "Beam proposals: {} ({} routes)",
        hero["name"].as_str().unwrap_or("hero"),
        proposals.len()
    );
    Ok(HeroProposals {
        proposals: proposals.into_values().collect(),
        diagnostics,
    })
}

fn nominate_search(
    database: &AnalysisDatabase,
    hero: &Value,
    data: &DiscoveryData,
    request: &SearchRequest<'_>,
    default: &Value,
    context: &ExportContext,
) -> Result<HeroProposals> {
    let result = search_group(request)?;
    let mut routes = result.routes;
    let default = core_items(default)?;
    routes.sort_by(|left, right| {
        right
            .score
            .total_cmp(&left.score)
            .then_with(|| {
                (left.core.iter().copied().collect::<BTreeSet<_>>() != default)
                    .cmp(&(right.core.iter().copied().collect::<BTreeSet<_>>() != default))
            })
            .then(right.owners.cmp(&left.owners))
            .then_with(|| left.core.cmp(&right.core))
            .then_with(|| left.path.cmp(&right.path))
    });
    let count = routes.len();
    let mut accepted = BTreeSet::new();
    let mut proposals = Vec::new();
    let mut rejected = BTreeMap::<String, u64>::new();
    for route in routes {
        if accepted.contains(&route.core) {
            continue;
        }
        match nominate_route(database, hero, data, &route, context) {
            Ok(row) => {
                accepted.insert(route.core);
                proposals.push(BeamNomination {
                    group: request.group.into(),
                    row,
                    scores: BTreeMap::from([(request.state.to_string(), route.score)]),
                });
            }
            Err(error) => *rejected.entry(error.to_string()).or_default() += 1,
        }
    }
    Ok(HeroProposals {
        proposals,
        diagnostics: vec![
            json!({"group":request.group,"state":request.state,"routes":count,"admitted_cores":accepted.len(),"rejections":rejected,"unassigned":result.unassigned,"expansions":result.expansions}),
        ],
    })
}

fn nominate_route(
    database: &AnalysisDatabase,
    hero: &Value,
    data: &DiscoveryData,
    route: &BeamRoute,
    context: &ExportContext,
) -> Result<Nomination> {
    let order = route
        .targets
        .iter()
        .filter(|item| route.core.contains(item))
        .copied()
        .collect::<Vec<_>>();
    let discovery_order = order_evidence(data, &route.core, &order, "discovery")?;
    let selection_order = order_evidence(data, &route.core, &order, "selection")?;
    if discovery_order["passes"] != true || selection_order["passes"] != true {
        return Err(Error::new(
            "Beam order lacks discovery or selection support",
        ));
    }
    let actions = replay_actions(&route.path, &route.core, &context.graph)?;
    if actions
        .last()
        .and_then(|step| step["cumulative_cost"].as_u64())
        != Some(route.spent)
    {
        return Err(Error::new(
            "Beam route differs from production mechanics replay",
        ));
    }
    let discovery = evaluate_core(data, &route.core, "discovery")?;
    let selection = evaluate_core(data, &route.core, "selection")?;
    let support = integer(&discovery, "owners")?;
    let reasons = SUPPORT.core_reasons(support, integer(&selection, "owners")?);
    if !reasons.is_empty() {
        return Err(Error::new(reasons.join("; ")));
    }
    let path = json!({"method":"beam16","order":order,"discovery":discovery_order,"selection":selection_order,"admitted_before_validation":true,"legal":true,"actions":actions,"reason":null});
    let names = route
        .core
        .iter()
        .map(|item| context.graph.require(*item).map(|node| node.name.clone()))
        .collect::<Result<Vec<_>>>()?;
    let cost = route
        .core
        .iter()
        .map(|item| context.graph.require(*item).map(|node| node.cost))
        .collect::<Result<Vec<_>>>()?
        .iter()
        .sum();
    let candidate = Candidate {
        items: route.core.clone(),
        names,
        cost,
        discovery_support: support,
        discovery_lift: deadlock_data::real(&discovery, "joint_lift")?,
        score: route.score,
        parent: Vec::new(),
        parent_retention: None,
        identity_id: core_identity(data.hero, &route.core),
        selection,
        selection_rejections: Vec::new(),
    };
    let guide = freeze_guide(
        database,
        data,
        &route.core,
        &order,
        &context.graph,
        Some(&route.path),
    )?;
    Ok(Nomination {
        candidate,
        hero_id: data.hero,
        selection_rank: 0,
        path,
        guide,
        tactics: describe_overlap(hero, &route.core, &context.normal_assets)?,
        branch_candidates: Vec::new(),
    })
}
